### ISSUE TITLE
NVME pass through

### ISSUE THREAD

**Issue by jorgefm** [Drive-Trust-Alliance/sedutil#463]
Hi all,

Is  there any possibility to have NVME pass through support implemented like for the SATA devices?

I've been trying to access to the OPAL info from a M.2 device connected through an ASMedia ASM2362 USB to NVMe bridge enclosure with no luck. After changing a little the 'DtaDevLinuxNvme::identify' method I've been able to access the information but I don't have more information to pass the NVME_SECURITY_RECV|SEND data in 'DtaDevLinuxNvme::sendCmd' embedded in a cdb structure.

To identify the device I've follow the 'https://www.smartmontools.org/ticket/1221' info.

`
void DtaDevLinuxNvme::identify(OPAL_DiskInfo& disk_info)
{
    LOG(D4) << "Entering DtaDevLinuxNvme::identify()";

    uint8_t ctrl[4096];
    sg_io_hdr_t sg;
    uint8_t sense[32]; // how big should this be??
    uint8_t cdb[16];
    memset(&cdb, 0, sizeof (cdb));
    memset(&sense, 0, sizeof (sense));
    memset(&sg, 0, sizeof (sg));

    // https://www.smartmontools.org/ticket/1221
    cdb[0] = 0xe6; // asmedia pass through
    cdb[1] = NVME_IDENTIFY;
    cdb[3] = 1;    // CDW10.byte0

    /*
     * Set up the SCSI Generic structure
     * see the SG HOWTO for the best info I could find
     */
    sg.interface_id = 'S';
    sg.dxfer_direction = SG_DXFER_FROM_DEV;
    sg.cmd_len = sizeof (cdb);
    sg.mx_sb_len = sizeof (sense);
    sg.iovec_count = 0;
    sg.dxfer_len = 4096;
    sg.dxferp = ctrl;
    sg.cmdp = cdb;
    sg.sbp = sense;
    sg.timeout = 60000;
    sg.flags = 0;
    sg.pack_id = 0;
    sg.usr_ptr = NULL;
    //    LOG(D4) << "cdb before ";
    //    IFLOG(D4) hexDump(cdb, sizeof (cdb));
    //    LOG(D4) << "sg before ";
    //    IFLOG(D4) hexDump(&sg, sizeof (sg));
    /*
     * Do the IO
     */
    if (ioctl(fd, SG_IO, &sg) < 0) {
        LOG(D4) << "cdb after ";
        IFLOG(D4) DtaHexDump(cdb, sizeof (cdb));
        LOG(D4) << "sense after ";
        IFLOG(D4) DtaHexDump(sense, sizeof (sense));
        disk_info.devType = DEVICE_TYPE_OTHER;
        return;
    }

    disk_info.devType = DEVICE_TYPE_NVME;
    uint8_t *results = ctrl;
    results += 4;
    memcpy(disk_info.serialNum, results, sizeof(disk_info.serialNum));
    results += sizeof(disk_info.serialNum);
    memcpy(disk_info.modelNum, results, sizeof(disk_info.modelNum));
    results += sizeof(disk_info.modelNum);
    memcpy(disk_info.firmwareRev, results, sizeof(disk_info.firmwareRev));

    return;
}`

**Comment by jorgefm** [comment]
Hi, thanks for the info!

Using the sedutil-cli against the M.2 device attached to the mainboard it works ok, but it's not very comfortable when you have to initialize several disks. Then, I was looking for a external drive to do that. I've tested several devices based in the Realtek RTL9210 and in the ASMedia ASM2362, but both are not supported in sedutil to be able to detect OPAL info and set it up.

The USB to SATA is working ok, but the USB to NVMe is a little behind. Maybe is needed some reverse engineering to capture the OPAL data when is working ok attached on mainboard but the hard point is how to map this info in a ASMedia pass through command without info about the ASM2362 device.

**Comment by hashkool** [comment]
Please let me know if I only add noise (I know little about this stuff). I see there is a patch for linux-**arm** which moves opal stuff from the pci layer to core: https://lore.kernel.org/linux-arm-kernel/20221108150252.2123727-3-hch@lst.de/#Z31drivers:nvme:host:pci.c

If you have an arm pc you could maybe see if you get different results?  I am not sure if these patches have been mainlined already btw.

**Comment by jorgefm** [comment]
Yes, the OPAL support for NVMe is already in the linux kernel and sedutil use it nicely... the problem is when you try to access this functionality through an intermediate USB to NVMe bridge. You must know how to map your request onto an ioctl the bridge (ASM2362 chip) is able to recognize and send it to the NVMe device attached... But I don't expect a lot from ASMedia to get the ASM2362 datasheet...

**Comment by hashkool** [comment]
I was under the impression that an unpatched kernel assumes that opal is tied to pci or sata, and that these code paths wouldn't be reached if using an usb bridge, as the kernel assumes it is talking to an usb disk. I thought the kernel patch would help in this situation, as it lifts the opal stuff from interface level to disk level. 

I think you know more about this stuff then me however.

**Comment by jorgefm** [comment]
Well, I've not analized the problem to the kernel level because I've got sedutil working when the NVME device is connected on the mainboard (the https://github.com/Drive-Trust-Alliance/sedutil/blob/master/linux/DtaDevLinuxNvme.cpp code path is used). 

When the device is attached to an enclosure with ASM2362 is detected has an SATA device and no info is showed about the device. Forcing to identify it using the code from the first message (extracted from smarttool tool) I was able to access to that information, then I guess that to get access to the rest of functionality I've to go the same way but I've not found information about it...

**Comment by hashkool** [comment]
Hmm, I am out of my depth here.

From this thread, it seems a bit of a hit and miss with adapters: https://forum.level1techs.com/t/nvme-to-usb-3-1-enclosure-buggy-in-linux-rtl9210b-chipset/199752

There are lots of drive specific quirks addressed on the pci level: https://github.com/torvalds/linux/commit/107b4e063d78c300b21e2d5291b1aa94c514ea5b.
I wonder 
1. if the ioctl will take [drivers/nvme/host/pci.c](https://github.com/torvalds/linux/commit/107b4e063d78c300b21e2d5291b1aa94c514ea5b#diff-7f40318de582e4a754803421ecf57f8a8009ada63ba1d956c49cc9275bd2c023) into account, and 
2. wether your adapter will faithfully pass through all info the kernel is poking for.

**Comment by hashkool** [comment]
Edit 2: Some adapters seem to be working, from a quick glance: https://github.com/Drive-Trust-Alliance/sedutil/issues/115

**Comment by Blacklands** [comment]
Has anyone been able to test an adapter with PCIe Tunneling (whether it's Thunderbolt 3 or fully-featured USB4/TB4)?
I've been hoping that those would be able to pass through the OPAL commands similar to a direct internal NVME -> PCIe connection.

I just got an enclosure with the [`Asmedia ASM2464PD`](https://www.asmedia.com.tw/product/802zX91Yw3tsFgm4/C64ZX59yu4sY1GW5) just to then find out that my only Thunderbolt 4 host device doesn't play nicely with it, and just falling back to USB Legacy Mode (and there the passthrough does _not_ work, sedutil doesn't detect the drive as an SED). :(
So I still haven't been able to answer this. 

Apart from that, I've only seen SATA -> USB adapters working, but there might be NVME -> USB ones out there too.

**Comment by JaBoMa** [comment]
I did not have problem with initialization of my Samsung 970 EVO Plus (M.2 NVMe) 1TB disk on my 12 Years old HP EliteBook 8540w, I was operating using DTA Sedutil Rescue System v.1.20 i.e. RESCUE32.img.gz from exec-1.20.0.zip or from exec-1.20.0.tar.gz that you can find under [https://github.com/Drive-Trust-Alliance/exec/tags](https://github.com/Drive-Trust-Alliance/exec/tags), and the SSD was connected via the AXAGON EEM2-UG2 USB-C to NVMe M.2 SSD interface. It was working well together. The rescue system was booted from USB memory Stick.
I was able to initialize that SSD as bootable TCG OPAL encrypted disk. The commands sedutil-cli --scan, sedutil-cli --query, and sedutil-cli --listLockingRanges <password> were also working well, 
The sedutil-cli.exe v. 1.20.0, that was started from Windows 7 on the same laptop, can't see that disk - connected the same way - at all.
I didn't try to reach that disk from the same laptop using sedutil-cli v.1.20.0 under Ubuntu.
I hope I did not add any noise to this thread. I was just checking possibilities to reach data from new SSD on other PC in the case the hosting PC fails.
ATB

**Comment by jorgefm** [comment]
> Edit 2: Some adapters seem to be working, from a quick glance: #115

Great! This issue has a lot of interesting info!! I'll try to get a RTL9210b adapter to test...

Thank you!
